### PR TITLE
Allow to post label with comments

### LINKED ISSUE #224: Allow specifying label when POSTing comments

### PR DESCRIPTION
**Pull request by stephawe**
Add a class NestedPKSerializer to serialize nested object
Add the label into the comment serializers
Add some tests

closes #224

### CODE REVIEW

File: democracy/views/utils.py [hunk 1]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+
+        id_ = value['id']
+        if not id_:
+            if self.required:
+                raise ValidationError(_('This field is required.'))
+            return None
+
+        return super().to_internal_value(id_)
+
+    def is_expanded(self):
```
Comment (akx) [inline democracy/views/utils.py]: Is this necessary? `expanded` will always be an attribute of a JSONPKRelatedField instance.

File: democracy/views/utils.py [hunk 2]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+    invalid_json_error = _('Incorrect JSON. Expected JSON, received %s.')
+    missing_id_error = _('The primary key is missing. Expected {"id": id, ...}, received %s.')
+
+    def __init__(self, *args, **kwargs):
+        self.related_serializer = kwargs.pop('serializer', None)
+        self.expanded = kwargs.pop('expanded', False)
+        super(JSONPKRelatedField, self).__init__(*args, **kwargs)
+
+    def use_pk_only_optimization(self):
+        if self.is_expanded():
```
Comment (akx) [inline democracy/views/utils.py]: `return not self.expanded` would do!

File: democracy/views/utils.py [hunk 3]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+            return False
+        else:
+            return True
+
+    def to_representation(self, obj):
+        if isinstance(self.related_serializer, str):
+            self.related_serializer = globals().get(self.related_serializer, None)
```
Comment (akx) [inline democracy/views/utils.py]: `globals().get()`? 😨 

Why is this sorcery required?

File: democracy/views/utils.py [hunk 4]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+        if self.is_expanded():
+            return self.related_serializer(obj, context=self.context).data
+        id_ = super(JSONPKRelatedField, self).to_representation(obj)
```
Comment (akx) [inline democracy/views/utils.py]: You can just use `id` even if it's a global name.

File: democracy/views/utils.py [hunk 5]
```diff
@@ -4,8 +4,10 @@
 from django.core.exceptions import ImproperlyConfigured
 from django.db.models.query import QuerySet
 from django.utils.timezone import now
+from django.utils.translation import ugettext as _
```
Comment (akx) [inline democracy/views/utils.py]: Should be `ugettext_lazy`.

File: democracy/views/utils.py [hunk 6]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+class JSONPKRelatedField(PrimaryKeyRelatedField):
+    """
+    Support of showing and saving of expanded JSON nesting or just a resource
+    ID.
+    Serializing is controlled by query string param 'expand', deserialization
+    by format of JSON given.
+    Default serializing is expand=false.
+    """
+
+    invalid_json_error = _('Incorrect JSON. Expected JSON, received %s.')
```
Comment (akx) [inline democracy/views/utils.py]: Should probably use the dict-y form of interpolation to make things clearer for translators:

> `received %(data)s`

File: democracy/views/utils.py [hunk 7]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+    missing_id_error = _('The primary key is missing. Expected {"id": id, ...}, received %s.')
+
+    def __init__(self, *args, **kwargs):
+        self.related_serializer = kwargs.pop('serializer', None)
+        self.expanded = kwargs.pop('expanded', False)
```
Comment (akx) [inline democracy/views/utils.py]: Maybe `self.expanded = bool(kwargs.pop('expanded', False))` to ensure booliness.

File: democracy/views/utils.py [hunk 8]
```diff
@@ -97,3 +99,56 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
         filters['%s__published' % hearing_lookup] = True
 
     return queryset.filter(**filters)
+
+
+class JSONPKRelatedField(PrimaryKeyRelatedField):
+    """
+    Support of showing and saving of expanded JSON nesting or just a resource
+    ID.
```
Comment (akx) [inline democracy/views/utils.py]: You can just move this `ID.` to the line above, wrapping rules be damned.

File: democracy/views/section_comment.py [hunk 9]
```diff
@@ -49,6 +57,13 @@ class SectionCommentSerializer(BaseCommentSerializer):
     """
     Serializer for comment added to section.
     """
+    label = JSONPKRelatedField(
+                queryset=Label.objects.all(),
```
Comment (akx) [inline democracy/views/section_comment.py]: Weird indentation. (The kwargs should be indented 8 spaces; the trailing `)` should be flush with `label = `.)

File: democracy/views/section_comment.py [hunk 10]
```diff
@@ -6,20 +6,28 @@
...
 from democracy.pagination import DefaultLimitPagination
-from democracy.views.utils import filter_by_hearing_visible
+from democracy.views.utils import filter_by_hearing_visible, JSONPKRelatedField
 
 
 class SectionCommentCreateSerializer(serializers.ModelSerializer):
     """
     Serializer for comments creation.
     """
+    label = JSONPKRelatedField(
```
Comment (akx) [inline democracy/views/section_comment.py]: Weird indentation. (The kwargs should be indented 8 spaces; the trailing `)` should be flush with `label = `.)

File: democracy/views/utils.py [hunk 11]
```diff
@@ -97,3 +99,47 @@ def filter_by_hearing_visible(queryset, request, hearing_lookup='hearing'):
...
+    def to_internal_value(self, value):
+        if not isinstance(value, dict):
+            raise ValidationError(self.invalid_json_error % {'data': type(value).__name__})
+        if 'id' not in value:
+            raise ValidationError(self.missing_id_error % {'data': value})
+
+        id = value['id']
+        if not id:
+            if self.required:
+                raise ValidationError(_('This field is required.'))
```
Comment (tuomas777) [inline democracy/views/utils.py]: Would be good if this specifies exactly which field is required, ie. raise `ValidationError({'id': [_('This field is required.')]})` or maybe use the `missing_id_error`?

**Comment by tuomas777** [comment]
I like this new field, and we will have a plenty of use for it in the future

**Comment by Rikuoja** [comment]
We will have use for it next week, if we manage to launch in time =)

**Comment by stephawe** [comment]
@akx: updated

**Pull request merged**